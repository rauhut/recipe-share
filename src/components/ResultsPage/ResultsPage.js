import React from "react";
import RecipeDisplay from "../RecipeDisplay/RecipeDisplay";

class ResultsPage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isLoading: true,
      recipes: [],
    };
  }

  async componentDidMount() {
    try {
      const res = await fetch(
        "https://recipe-share-backend.herokuapp.com/recipes",
        {
          method: "get",
          headers: {
            "Content-Type": "application/json",
          },
        }
      );
      if (!res.ok) {
        throw new Error("Something went wrong when trying to get recipes");
      }
      const recipes = await res.json();
      this.setState({ recipes: recipes.data });
      this.setState({ isLoading: false });
    } catch (err) {
      console.log(err);
    }
  }

  filterRecipes = () => {
    return this.state.recipes.filter((recipe) => {
      return recipe.name
        .toLowerCase()
        .includes(this.props.searchEntry.toLowerCase());
    });
  };

  render() {
    return (
      <div>
        <h1>Search Results for "{this.props.searchEntry}"</h1>
        <RecipeDisplay recipes={this.filterRecipes()} />
      </div>
    );
  }
}

export default ResultsPage;
