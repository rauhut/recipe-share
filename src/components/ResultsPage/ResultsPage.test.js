import React from "react";
import { shallow } from "enzyme";
import ResultsPage from "./ResultsPage";

describe("ResultsPage", () => {
  let wrapper;

  const mockProps = {
    searchEntry: "test",
  };

  const mockRecipes = [
    {
      name: "cookies",
      ingredients: ["flour", "sugar", "butter", "salt", "chocolate"],
      steps: ["step 1", "step 2", "step 3"],
      cookTime: "30 mins",
      description: "just basic cookies",
    },
    {
      name: "meatballs",
      ingredients: ["ground beef", "eggs", "bread crumbs", "garlic", "onion"],
      steps: ["step 1", "step 2", "step 3"],
      cookTime: "30 mins",
      description: "just basic meatballs",
    },
  ];

  it("renders page with resultswithout crashing", () => {
    wrapper = shallow(<ResultsPage {...mockProps} />);
    wrapper.setState({ recipes: mockRecipes });
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("displays the recipes and hide loading spinner on api call success", (done) => {
    global.fetch = jest.fn();

    wrapper = shallow(<ResultsPage {...mockProps} />, {
      disableLifecycleMethods: true,
    });

    const spyDidMount = jest.spyOn(ResultsPage.prototype, "componentDidMount");

    fetch.mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        json: () => {
          return Promise.resolve({
            data: mockRecipes,
          });
        },
      });
    });

    const didMount = wrapper.instance().componentDidMount();

    expect(spyDidMount).toHaveBeenCalled();

    didMount.then(() => {
      wrapper.update();
      expect(wrapper.state("recipes")).toEqual(mockRecipes);
      expect(wrapper.state("isLoading")).toEqual(false);
      spyDidMount.mockRestore();
      fetch.mockClear();
      done();
    });
  });

  it("filtered recipes correctly- one result", () => {
    const mockPropsOneResult = {
      searchEntry: "cookies",
    };
    wrapper = shallow(<ResultsPage {...mockPropsOneResult} />);
    wrapper.setState({ recipes: mockRecipes });
    expect(wrapper.instance().filterRecipes()).toEqual([
      {
        name: "cookies",
        ingredients: ["flour", "sugar", "butter", "salt", "chocolate"],
        steps: ["step 1", "step 2", "step 3"],
        cookTime: "30 mins",
        description: "just basic cookies",
      },
    ]);
  });

  it("filtered recipes correctly- no result", () => {
    const mockPropsOneResult = {
      searchEntry: "z",
    };
    wrapper = shallow(<ResultsPage {...mockPropsOneResult} />);
    wrapper.setState({ recipes: mockRecipes });
    expect(wrapper.instance().filterRecipes()).toEqual([]);
  });
});
