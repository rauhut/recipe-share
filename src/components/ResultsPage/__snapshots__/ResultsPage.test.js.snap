// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsPage renders page with resultswithout crashing 1`] = `
"<div>
  <h1>
    Search Results for &quot;
    test
    &quot;
  </h1>
  <RecipeDisplay recipes={{...}} />
</div>"
`;
